### PR TITLE
Ruby: CFG: make all expressions "post-order" nodes

### PR DESCRIPTION
**Pull request by aibaars**

### CODE REVIEW

File: ruby/ql/lib/codeql/ruby/controlflow/internal/ControlFlowGraphImpl.qll
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/controlflow/internal/ControlFlowGraphImpl.qll:401]: Don't we need a `c instanceof NormalCompletion` check here?

File: ruby/ql/lib/codeql/ruby/controlflow/internal/ControlFlowGraphImpl.qll
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/controlflow/internal/ControlFlowGraphImpl.qll:450]: Move up after `succ = this and` to also cover the first disjunct.

File: ruby/ql/lib/codeql/ruby/controlflow/internal/ControlFlowGraphImpl.qll
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/controlflow/internal/ControlFlowGraphImpl.qll:772]: I think this should be `c.(MatchingCompletion).getValue() = true`.

File: ruby/ql/lib/codeql/ruby/controlflow/internal/ControlFlowGraphImpl.qll
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/controlflow/internal/ControlFlowGraphImpl.qll:1164]: add `c instanceof NormalCompletion`

**Review by nickrolfe — COMMENTED** [review]
I haven't reviewed the test output, but the changes in `ControlFlowGraphImpl.qll` all look sensible.

**Review by hvitved — CHANGES_REQUESTED** [review]
Looks great, it even simplified some existing logic. I suggest we rebase this once https://github.com/github/codeql/pull/7393 has been merged, and then get rid of the last commit.

File: ruby/ql/lib/codeql/ruby/controlflow/internal/ControlFlowGraphImpl.qll
Comment (aibaars) [reply ruby/ql/lib/codeql/ruby/controlflow/internal/ControlFlowGraphImpl.qll:772]: A pattern may also end with a simple completion, for example a `*var` never fails.

File: ruby/ql/lib/codeql/ruby/controlflow/internal/ControlFlowGraphImpl.qll
Comment (aibaars) [reply ruby/ql/lib/codeql/ruby/controlflow/internal/ControlFlowGraphImpl.qll:401]: don't we get that "for free" because `value` and `branch(_)` are mentioned in `propagatesAbnormal`?

File: ruby/ql/lib/codeql/ruby/controlflow/internal/ControlFlowGraphImpl.qll
Comment (aibaars) [reply ruby/ql/lib/codeql/ruby/controlflow/internal/ControlFlowGraphImpl.qll:773]: Could write the following instead `c instanceof MatchingCompletion implies c.(MatchingCompletion).getValue() = true`

File: ruby/ql/consistency-queries/CfgConsistency.ql
Comment (aibaars) [reply ruby/ql/consistency-queries/CfgConsistency.ql]: That gives quite a few results : https://github.com/github/codeql/runs/4549392461?check_suite_focus=true

Comment (hvitved) [reply ruby/ql/consistency-queries/CfgConsistency.ql:8]: That made me realize that the modelling of `InClause` is actually not like (I think) it should be. There is a subtle difference: before, the node for the `InClause` itself was visited regardless of a match, but now it is only visited if there is a match.

There are two options: Revert 1343ed58a21eec2954876d8d42e877a382ba89c8 or use CFG splitting to make sure that `InClauses` are truly post-order. I don't think splitting is worth the effort, so I suggest a revert.

With a revert, we can update the check above to

```ql
query predicate nonPostOrderExpr(Expr e, string cls) {
  cls = e.getPrimaryQlClasses() and
  not exists(e.getDesugared()) and
  not e instanceof InClause and
  not e instanceof BeginExpr and
  not e instanceof Namespace and
  not e instanceof Toplevel and
  exists(AstNode last, Completion c |
    last(e, last, c) and
    last != e and
    c instanceof NormalCompletion
  )
}
```

File: ruby/ql/lib/codeql/ruby/controlflow/internal/ControlFlowGraphImpl.qll
Comment (hvitved) [reply ruby/ql/lib/codeql/ruby/controlflow/internal/ControlFlowGraphImpl.qll:401]: `propagatesAbnormal` only gives you the escaping `raise` edge (typically to the exit point), it doesn't prevent an internal edge.

File: ruby/ql/consistency-queries/CfgConsistency.ql
Comment (aibaars) [reply ruby/ql/consistency-queries/CfgConsistency.ql:8]: > That made me realize that the modelling of `InClause` is actually not like (I think) it should be. There is a subtle difference: before, the node for the `InClause` itself was visited regardless of a match, but now it is only visited if there is a match.

I think this is the correct behaviour; if there is no match then the `InClause` has no value (not even `nil`).

Comment (aibaars) [reply ruby/ql/consistency-queries/CfgConsistency.ql:8]: As discussed, perhaps `InClause` shouldn't be an `Expr` in the first place.

**Review by hvitved — APPROVED** [review]
LGTM. I have started a final DCA run; let's wait for that before we merge.

**Review by hvitved — DISMISSED** [review]

**Review by hvitved — APPROVED** [review]

**Pull request merged**
